File: app/src/main/java/me/devsaki/hentoid/notification/update/UpdateAvailableNotification.java
package me.devsaki.hentoid.notification.update;

import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import androidx.annotation.NonNull;
import androidx.core.app.NotificationCompat;

import me.devsaki.hentoid.R;
import me.devsaki.hentoid.services.UpdateDownloadService;
import me.devsaki.hentoid.util.PendingIntentCompat;
import me.devsaki.hentoid.util.notification.Notification;

public class UpdateAvailableNotification implements Notification {

    private final String updateUrl;

    public UpdateAvailableNotification(String updateUrl) {
        this.updateUrl = updateUrl;
    }

    @NonNull
    @Override
    public android.app.Notification onCreateNotification(Context context) {
        Intent intent = UpdateDownloadService.makeIntent(context, updateUrl);

        PendingIntent pendingIntent = PendingIntentCompat.getForegroundService(context, intent);

        return new NotificationCompat.Builder(context, UpdateNotificationChannel.ID)
                .setSmallIcon(R.drawable.ic_stat_hentoid)
                .setPriority(NotificationCompat.PRIORITY_MAX)
                .setCategory(NotificationCompat.CATEGORY_MESSAGE)
                .setVibrate(new long[]{1, 1, 1})
                .setContentTitle("An update is available!")
                .setContentText("Tap to download")
                .setContentIntent(pendingIntent)
                .build();
    }
}

File: app/src/main/java/me/devsaki/hentoid/notification/update/UpdateAvailableNotification.kt
package me.devsaki.hentoid.notification.update

import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.os.Build
import androidx.core.app.NotificationCompat
import me.devsaki.hentoid.R
import me.devsaki.hentoid.receiver.AppUpdateDownloadReceiver
import me.devsaki.hentoid.util.notification.Notification
import me.devsaki.hentoid.workers.data.UpdateDownloadData

class UpdateAvailableNotification(private val downloadUrl: String) : Notification {

    override fun onCreateNotification(context: Context): android.app.Notification {
        val intent = Intent(context, AppUpdateDownloadReceiver::class.java)
        val builder = UpdateDownloadData.Builder().setUrl(downloadUrl)
        intent.putExtras(builder.bundle)
        val flags =
            if (Build.VERSION.SDK_INT > 30)
                PendingIntent.FLAG_CANCEL_CURRENT or PendingIntent.FLAG_IMMUTABLE
            else PendingIntent.FLAG_CANCEL_CURRENT
        val pendingIntent =
            PendingIntent.getBroadcast(context, 0, intent, flags)

        return NotificationCompat.Builder(context, UpdateNotificationChannel.ID)
            .setSmallIcon(R.drawable.ic_hentoid_shape)
            .setPriority(NotificationCompat.PRIORITY_MAX)
            .setCategory(NotificationCompat.CATEGORY_MESSAGE)
            .setVibrate(longArrayOf(1, 1, 1))
            .setContentTitle(context.resources.getText(R.string.update_available))
            .setContentText(context.resources.getText(R.string.tap_to_download))
            .setContentIntent(pendingIntent)
            .setAutoCancel(true)
            .build()
    }
}
